### DISCUSSION TITLE
Support for "getConstraintJacobianTimeDerivative"

### DISCUSSION THREAD

**Discussion by Cfather** [stack-of-tasks/pinocchio#2442] [Q&A] [answered]
Hi, I have a few questions related to general constraints representation in Pinocchio.

1. Is there any paper that I could refer to in order to understand the math in [constrained-dynamics.hpp](https://github.com/stack-of-tasks/pinocchio/blob/master/include/pinocchio/algorithm/constrained-dynamics.hpp)?
2. There have been multiple technical routes to solve dynamics in the presence of contacts. Different simulators, such as MuJoCo, Pybullet, Dojo, and DART have different theories and implementations. What would be the roadmap for pinocchio in terms of contact solver?
3. I wonder if there will be support for "getConstraintJacobianTimeDerivative", which is the time derivative of "getConstraintJacobian" in [contact-jacobian.hpp](https://github.com/stack-of-tasks/pinocchio/blob/master/include/pinocchio/algorithm/contact-jacobian.hpp). I thought the jacobian time derivative was useful to enforce constraints on the second order. $\frac{d}{dt}(J(q)v) = 0 \rightarrow \frac{d}{dt}(J(q)) v + J(q)a = 0$ and I'm surprised that this is not used in solving contact dynamics currently. I guess this might be related to the answer to question 2.

Looking forward to any answers! Thanks!

**Comment by jcarpent** [comment, marked as answer]
Yes, of course @Cfather:

> 1. Is there any paper that I could refer to in order to understand the math in [constrained-dynamics.hpp](https://github.com/stack-of-tasks/pinocchio/blob/master/include/pinocchio/algorithm/constrained-dynamics.hpp)?

You can read this https://inria.hal.science/hal-03271811/document

> 2\. There have been multiple technical routes to solve dynamics in the presence of contacts. Different simulators, such as MuJoCo, Pybullet, Dojo, and DART have different theories and implementations. What would be the roadmap for pinocchio in terms of contact solver?

Together with @quentinll and @lmontaut, we have been working on more robust, efficient, and scalable solvers for frictional dynamics in robotics, able to handle rigid and compliant contact models. I invite you to read https://hal.science/hal-04588906v1/file/simplecontacts2024.pdf


> 3\. I wonder if there will be support for "getConstraintJacobianTimeDerivative", which is the time derivative of "getConstraintJacobian" in [contact-jacobian.hpp](https://github.com/stack-of-tasks/pinocchio/blob/master/include/pinocchio/algorithm/contact-jacobian.hpp). I thought the jacobian time derivative was useful to enforce constraints on the second order. 
>          and I'm surprised that this is not used in solving contact dynamics currently. I guess this might be related to the answer to question 2.

When you work with frictionnal formulations, you have to work with impulses, no more question of joint acceleration.
We have started to work on fully diffrentiable simulator (with SOTA timings to the best of our knowledge). Please have a look to:
https://hal.science/hal-04694092v1/file/simple2024gradients.pdf

**Reply by Cfather** [reply]
Thank you very much for your response! Can't wait to see these new features!!!

**Reply by Cfather** [reply]
May I ask if these features have been incorporated into the master branch of pinocchio already? Where can I find the implementation of these methods?

**Reply by jcarpent** [reply]
The two first points have been already impletemented. Search for ADMMSolver and you will find the useful part ;)

**Reply by Cfather** [reply]
Great. Thank you!
